eventlog/collect_txt_event.go

```diff
@@ -0,0 +1,141 @@
+	}
+
+	// Parse the txt-event log data
+	biosDataSize := binary.LittleEndian.Uint64(mmap[0:])
+	osMleDataSize := binary.LittleEndian.Uint64(mmap[biosDataSize:])
+	if osMleDataSize <= 0 {
+		return errors.New("eventlog/collect_txt_event:updateTxtEventLog() Invalid osMleDataSize")
+	}
+
+	// Read OsSinitData (Table 22. OS to SINIT Data Table) at HeapBase+BiosDataSize+OsMleDataSize+8
+	osSinitVersion := binary.LittleEndian.Uint32(mmap[biosDataSize+osMleDataSize+Uint64Size:])
+	if osSinitVersion >= 6 {
+		log.Infof("eventlog/collect_txt_event:updateTxtEventLog() OSInitData.Version = %d", osSinitVersion)
+	} else {
+		return errors.New("eventlog/collect_txt_event:updateTxtEventLog() OSInitData.Version was less than 6")
+	}
+
+	// ExtDataElement that is HEAP_EVENT_LOG_POINTER_ELEMENT2_1. ie OsSinitData.ExtDataElements[0].Type must be 0x8.
+	osSinitExtType := binary.LittleEndian.Uint32(mmap[biosDataSize+osMleDataSize+Uint64Size+ExtDataElementOffset:])
+	if osSinitExtType != 0x8 {
+		return errors.New("eventlog/collect_txt_event:updateTxtEventLog() OsSinitData.ExtDataElements[0].Type was not 0x8")
+	}
+
+	// Data is parsed based on HEAP_EVENT_LOG_POINTER_ELEMENT2_1 of Intel TXT spec 16.2. Reading EventLogPointer (20 bytes)
+	physicalAddress := binary.LittleEndian.Uint64(mmap[biosDataSize+osMleDataSize+Uint64Size+ExtDataElementOffset+Uint64Size:])
+	allocatedEventContainerSize := binary.LittleEndian.Uint32(mmap[biosDataSize+osMleDataSize+Uint64Size+ExtDataElementOffset+Uint64Size+Uint64Size:])
+	firstRecordOffset := binary.LittleEndian.Uint32(mmap[biosDataSize+osMleDataSize+Uint64Size+ExtDataElementOffset+Uint64Size+Uint64Size+Uint32Size:])
+	nextRecordOffset := binary.LittleEndian.Uint32(mmap[biosDataSize+osMleDataSize+Uint64Size+ExtDataElementOffset+Uint64Size+Uint64Size+Uint32Size+Uint32Size:])
+	firstEventLogOffset := (physicalAddress - eventLog.TxtHeapBaseAddr) + uint64(firstRecordOffset)
+	firstEventLogBuffer := bytes.NewBuffer(mmap[firstEventLogOffset:])
+
+	// Skip TCG_PCR_EVENT(Intel TXT spec. ver. 16.2) log data and put the remaining eventlog in measure-log.json
+	err = parseEventLog(firstEventLogBuffer, allocatedEventContainerSize)
+	if err != nil {
+		return errors.Wrap(err, "eventlog/collect_txt_event:updateTxtEventLog() Error while parsing EventLog data")
+	}
+
+	// Parse eventlog from nextRecordOffset and put in measure-log.json
+	if nextRecordOffset != 0 {
+		nextEventLogOffset := (physicalAddress - eventLog.TxtHeapBaseAddr) + uint64(nextRecordOffset)
+		nextEventLogBuffer := bytes.NewBuffer(mmap[nextEventLogOffset:])
+		err = createMeasureLog(nextEventLogBuffer, allocatedEventContainerSize-uint32(nextEventLogOffset))
+		if err != nil {
+			return errors.Wrap(err, "eventlog/collect_txt_event:updateTxtEventLog() Error while parsing nextEventLogOffset data")
+		}
+	}
+
+	// Unmap the /dev/mem buffer
+	err = syscall.Munmap(mmap)
```

Comment:
What will happen if an error condition occurs before this call to Munmap?  Should this be in a "defer" statement after the mmap is created?